### PR TITLE
azure service fabric service discovery

### PR DESCRIPTION
**Pull request by JavierZunzunegui**
Adding service discovery support for [Azure Service Fabric](https://docs.microsoft.com/en-us/azure/service-fabric/)

Uses the config `AzureServiceFabricSDConfig` to find [registered endpoints](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-service-manifest-resources#endpoints) contained in 
`Endpoints            map[string]struct{}`
and scrapes them. 

If monitoring a secure cluster, client authentication requires `ClientCertAndKeyFile` with a [client certificate](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-security-update-certs-azure#add-a-secondary-cluster-certificate-using-the-portal). Certificates with read-only access are enough and recommended.

Does not support discovery of containers yet.

Example scrape config: 

```
scrape_configs:
- job_name: service_fabric
  azure_sf_sd_configs:
  - cluster_url: "127.0.0.1:19080" # set to the desired cluster
    endpoints: {'Debug'}
    client_cert_and_key_file: "read-only.crt,read-only.key" # path to certs

  relabel_configs:
  - source_labels: ['__meta_azure_sf_service']
    regex:         "(.*)/(.*)"
    target_label:  "service"
    replacement:   "${1}"
  - source_labels: ['__meta_azure_sf_service']
    regex:         "(.*)/(.*)"
    target_label:  "component"
replacement: "${2}"
```

@brian-brazil

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -1435,3 +1443,26 @@ func (c *RemoteReadConfig) UnmarshalYAML(unmarshal func(interface{}) error) erro
 	}
 	return nil
 }
+
+// AzureServiceFabricSDConfig is the configuration for Azure Service Fabric service discovery.
+type AzureServiceFabricSDConfig struct {
```
Comment (brian-brazil) [inline config/config.go]: This doesn't look like the existing azure SD, why is it doing auth etc. differently?

File: config/config.go [hunk 2]
```diff
@@ -1435,3 +1443,26 @@ func (c *RemoteReadConfig) UnmarshalYAML(unmarshal func(interface{}) error) erro
...
+	ClusterHost          string              `yaml:"cluster_host"`
+	Endpoints            map[string]struct{} `yaml:"endpoints"`
```
Comment (brian-brazil) [inline config/config.go]: Pull down everything

File: config/config.go [hunk 3]
```diff
@@ -1435,3 +1443,26 @@ func (c *RemoteReadConfig) UnmarshalYAML(unmarshal func(interface{}) error) erro
...
+	RefreshInterval      model.Duration      `yaml:"refresh_interval,omitempty"`
+	ClientCertAndKeyFile string              `yaml:"client_cert_and_key_file"` // populate only for secure clusters. Comma separated
```
Comment (brian-brazil) [inline config/config.go]: Use the existing way of doing this rather than reimplementing this yourself.

**Comment by JavierZunzunegui** [comment]
The current Azure SD discovers Azure VMs. Service Fabric relies on [Virtual Machine Scale Sets](https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/).

Azure VMSS have been mentioned in this [issue](https://github.com/prometheus/prometheus/issues/2369)

**Comment by JavierZunzunegui** [comment]
To clarify, the existing Azure VM service discovery doesn't help in discovering Service Fabric applications (SF uses VMSS not VM). 
Discovery for Service Fabric here proposed is not happening at the level of VMSS but instead using [Service Fabric's REST API](https://docs.microsoft.com/en-us/rest/api/servicefabric/sfclient-index)

**Comment by JavierZunzunegui** [comment]
Also discovering through Service Fabric REST API has no quota limits (requests hit your own cluster) while comparable Azure API calls do have limits

File: config/config.go (hunk 1 above)
Comment (JavierZunzunegui) [inline config/config.go]: It is doing discovery of Service Fabric applications via the Service Fabric REST API, as opposed to discovery of Azure VMs as the existing Azure discovery mechanism does.

**Comment by krasi-georgiev** [comment]
due to lack of man power to review and maintain new SD we are on permanent hold for new SD.

As a replacement we have now added an example how to implement a custom SD adapter that converts any SD to json file that can be used by the file SD 
https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd

**Pull request closed without merging**